src/types/connection/connection_type.rs

```diff
@@ -1,97 +1,136 @@
-use crate::types::connection::cursor::Cursor;
-use crate::types::connection::edge::Edge;
-use crate::types::connection::page_info::PageInfo;
+use crate::connection::edge::Edge;
+use crate::connection::page_info::PageInfo;
+use crate::types::connection::{CursorType, EmptyEdgeFields};
 use crate::{
-    do_resolve, registry, Context, ContextSelectionSet, EmptyEdgeFields, Error, ObjectType,
+    do_resolve, registry, Context, ContextSelectionSet, Error, FieldResult, ObjectType,
     OutputValueType, Positioned, QueryError, Result, Type,
 };
 use async_graphql_parser::query::Field;
+use futures::{Stream, TryStreamExt};
 use indexmap::map::IndexMap;
 use inflector::Inflector;
 use itertools::Itertools;
 use std::borrow::Cow;
 
+/// The record type output by the data source
+pub struct Record<C, E, T> {
+    cursor: C,
+    edge: E,
+    element: T,
+}
+
+impl<C, E, T> Record<C, E, T> {
+    pub fn new(cursor: C, edge: E, element: T) -> Self {
+        Self {
+            cursor,
+            edge,
+            element,
+        }
+    }
+}
+
+impl<C, T> Record<C, EmptyEdgeFields, T> {
+    pub fn new_without_edge_fields(cursor: C, element: T) -> Self {
```

Comment:
I think we should make `new` be without EdgeFields and the make this `new_with_edge_fields`. The right defaults are important.

Reply:
If we change the name of my suggestion above, we could call this `new_with_additional_fields`.